File: lintcode/rotate_string/py/rotate_string.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/rotate-string
@Language: Python
@Datetime: 17-02-15 15:07
'''

class Solution:
    # @param s: a list of char
    # @param offset: an integer 
    # @return: nothing
    def rotateString(self, s, offset):
        if len(s) == 0:
            return
        
        k     = offset % len(s)
        s[:]  = s[::-1]
        s[:k] = reversed(s[:k])
        s[k:] = reversed(s[k:])

File: lintcode/easy/rotate_string/cpp/rotate_string.cpp
/*
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/rotate-string
@Language: C++
@Datetime: 17-02-15 19:16
*/

#include <algorithm>

class Solution {
public:
    /**
     * @param str: a string
     * @param offset: an integer
     * @return: nothing
     */
    void rotateString(std::string &str,int offset) 
    {
        if (str.size() == 0) {
            return;
        }
        
        const auto k = offset % str.size();
        std::reverse(str.begin(), str.end());
        std::reverse(str.begin(), str.begin() + k);
        std::reverse(str.begin() + k, str.end());
    }
};
